File: temp/tweet-test.scala
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf
import scala.util.parsing.json._
//import java.util.Calendar
import java.text.SimpleDateFormat
import java.util.Date

/**
def convert_to_1hr(timestamp: Long): String = {
   
   val min30 = timestamp.slice(11,13).toInt/30*30
   timestamp.take(11) + f"${min30}%02d" + "00"
}
*/
def convert_to_1hr(time: Date): String = {
   val hourFormat = new SimpleDateFormat("yyyyMMddHH")
   hourFormat.format(time)
}
/**
def matchTweet(json_line: String): String = x match {
}
*/
//val tweet = JSON.parseFull("""{"text": "@ImYourEpiphany \ud83d\ude02\ud83d\ude02\ud83d\ude02 no. Not at all.", "coordinates": [], "timestamp_ms": "1442469035685", "entities": {"user_mentions": ["ImYourEpiphany","ZafulUs"], "hashtags": ["Zaful","THashTag"]}, "place": {"full_name": "Bell, CA", "country_code": "US", "name": "Bell", "place_type": "city"}, "user": {"id_str": "3271546344", "location": "Los Angeles, CA"}, "id_str": "644387956707295233"}""")

//val conf = new SparkConf().setAppName("PriceDataExercise")
//val sc = new SparkContext(conf)
//val fileIn = "hdfs://ec2-54-209-187-157.compute-1.amazonaws.com:9000/user/test-tweets/json-tweets-10.txt"
//val fileOut = "hdfs://ec2-54-209-187-157.compute-1.amazonaws.com:9000/user/test-tweets/jsons/json-tweets-10.out2"
val fileIn = "TrendingTopics/temp/jsons/json-tweets-10.txt"
val fileOut = "TrendingTopics/temp/jsons/json-tweets-10.out2"
val file = sc.textFile(fileIn)
val trends = file.map(line => {
// println("before a tweet")
// println(json_tweet)
// println("after a tweet")

 val json_tweet = JSON.parseFull(line)
 json_tweet match {
  case Some(t: Map[String, _]) => {
    val t_id = t("id_str") // String / Long
    val text = t("text") // String — the tweet
    val user = t("user") // Map
    val coords = t("coordinates") // List

    val time_ms_str = t("timestamp_ms").toString // String 
    val time_ms : Long = time_ms_str.toLong // Long
    val date = new Date(time_ms)
    val hourSlot = convert_to_1hr(date)

    val place = t("place") // Map
    val p = place.asInstanceOf[Map[String,String]]
    val city = p("name")
    val country = p("country_code")

    val entities = t("entities") // Map
    val e = entities.asInstanceOf[Map[String,List[String]]]
    val userMentions = e("user_mentions")
    val hashtags = e("hashtags")


    // ((time, (cc,city)), (tag, 1)) (id?)
    val trendingUsers = userMentions.map(user => {
				((hourSlot, (country, city), user), 1)
			})
    val trendingTags = hashtags.map(tag => {
				((hourSlot, (country, city), tag), 1)
			})

    val trendingItems = trendingTags ::: trendingUsers // combine both

    

//    println(trendingItems)
    
//    (t_id, place._1, place._2)
  } // case  t:
  case _ => ((0, (0,0),0), 1)
 }
//}).reduceByKey(_+_).flatten

})
print(trends)
trends.saveAsTextFile(fileOut)

File: temp/tweet-test.py
from pyspark import SparkContext, SparkConf
import json
import sys
import datetime

hdfs = "hdfs://ec2-54-209-187-157.compute-1.amazonaws.com:9000"
path = "/user/test-tweets/tweets-17-22.json"
#path = "/user/test-tweets/json-tweets-10.txt"
# can take the path to a folder/file as an argument to be added to the hdfs path
if len(sys.argv) > 1:
    path = sys.argv[1]

folder_in = hdfs + path

# take a datetime object and return a string for the minute slot
def convert_to_1m(dt):
    return dt.strftime('%Y%m%d%H%M')

# take a datetime object and return a string for the hour slot
def convert_to_1h(dt):
    return dt.strftime('%Y%m%d%H')

# take a datetime object and return a string for the day slot
def convert_to_1d(dt):
    return dt.strftime('%Y%m%d')

# get poster's user
def userId(item):
    if ('user' in item) and isinstance(item['user'], dict):
	if 'id_str' in item['user']:
	    return item['user']['id_str']
    return None

# get a combined list of the hashtags and user mentions 
def trendItems(item):
    tags = []
    if ('entities' in item) and isinstance(item['entities'], dict):
	# only continue if we have a list (could be 'None')
	if 'hashtags' in item['entities']:
	    tags.extend(item['entities']['hashtags'])

	if 'user_mentions' in item['entities']:
	    tags.extend(item['entities']['user_mentions'])
    return tags

def tweetCountry(item):
    if ('place' in item) and isinstance(item['place'], dict):
        if 'country_code' in item['place']:
	    return item['place']['country_code']
    return None

def tweetCity(item):
    if ('place' in item) and isinstance(item['place'], dict):
        if 'name' in item['place']:
	    return item['place']['name']
    return None

# parse the next raw text line into a tweet -- just the needed elements
def tweet_from_json_line(json_line):
    try:
        item = json.loads(json_line)
    except:
	return None # not a json line -- just retun None

    if ('id_str' not in item) or ('timestamp_ms' not in item) or('text' not in item):
	return None # not what we expected

    tweet = {}
    tweet['t_id'] = item['id_str']
    tweet['u_id'] = userId(item) # just the userId is enough
    tweet['coors'] = item['coordinates'] # list [latitude,longitude]
    tweet['tags'] = trendItems(item) # will have the combined list of hashtags and user mentions
    tweet['city'] = tweetCity(item)
    tweet['country'] = tweetCountry(item) # the country code: US, UK, etc.

    # time related fields
    tweet['time_ms'] = item['timestamp_ms']
    tweetTime = datetime.datetime.fromtimestamp(long(tweet['time_ms'])/1000)
    tweet['minuteSlot'] = convert_to_1m(tweetTime)
    tweet['hourSlot'] = convert_to_1h(tweetTime)
    tweet['daySlot'] = convert_to_1d(tweetTime)

#    tweet['text'] = item['text'] # don't need the actual tweet...

    return tweet


# main work...
conf = (SparkConf().setAppName("Tweets-Py-mvp"))
sc = SparkContext(conf = conf)
tweet_jsons = sc.textFile(folder_in)

tweets = tweet_jsons.map(tweet_from_json_line)#.persist

tweetsWithTags = tweets.filter(lambda tweet : (tweet is not None) and ('tags' in tweet) and (len(tweet['tags']) > 0))

tagsAsValue = tweetsWithTags.map(lambda tweet : ((tweet['daySlot'], tweet['hourSlot'], tweet['minuteSlot'], tweet['country'], tweet['city']), tweet['tags']))
singleTagTuples = tagsAsValue.flatMapValues(lambda x : x)#.persist # ((daySlot, hourSlot, minuteSlot, country, city), tag)

tagPlaceDailyCount = singleTagTuples.map(lambda ((d, h, m, cc, c), t) : ((d, cc, c, t), 1)).reduceByKey(lambda x, y: x + y) 
tagPlaceHourlyCount = singleTagTuples.map(lambda ((d, h, m, cc, c), t) : ((h, cc, c, t), 1)).reduceByKey(lambda x, y: x + y) 
tagPlaceMinuteCount = singleTagTuples.map(lambda ((d, h, m, cc, c), t) : ((m, cc, c, t), 1)).reduceByKey(lambda x, y: x + y) 
	
folder_out = folder_in + "_pyOut"
tagPlaceDailyCount.saveAsTextFile(folder_out+"D")
tagPlaceHourlyCount.saveAsTextFile(folder_out+"H")
tagPlaceMinuteCount.saveAsTextFile(folder_out+"M")
#singleTagTuples.saveAsTextFile(folder_out)
